model_signing/signing/in_toto.py

```diff
@@ -0,0 +1,125 @@
+# Copyright 2024 The Sigstore Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Signing payloads for models as in-toto statements.
+
+To generate the signing payload we convert model manifests to in-toto formats,
+as described by https://github.com/in-toto/attestation/tree/main/spec/v1. The
+envelope format is DSSE, see https://github.com/secure-systems-lab/dsse.
+"""
+
+from typing import Final, Self
+from typing_extensions import override
+
+from in_toto_attestation.v1 import statement
+
+from model_signing.manifest import manifest as manifest_module
+from model_signing.signing import signing
+
+
+class IntotoPayload(signing.SigningPayload):
+    """A generic payload in in-toto format.
+
+    This class is abstract for now as we will support multiple payload formats
+    below.
+
+    Each subclass defines a constant for the predicate type class attribute
+    defined below.
+    """
+
+    predicate_type: Final[str]
+
+
+class SingleDigestIntotoPayload(IntotoPayload):
+    """In-toto payload where the model is serialized to just one digest.
+
+    In this case, we encode the model as the only subject of the statement. We
+    don't set the name field, and use the digest as the one resulting from the
+    model serialization.
+
+    However, since we use custom hashing algorithms, but these are not supported
+    by existing tools, we claim that the digest algorithm is sha-256 and include
+    the real digest in the predicate.
+
+    Example:
+    ```json
+    {
+      "_type": "https://in-toto.io/Statement/v1",
+      "subject": [
+        {
+          "digest": {
+            "sha256": "3aab065c...."
+          }
+        }
+      ],
+      "predicateType": "https://model_signing/Digest/v0.1",
+      "predicate": {
+        "actual_hash_algorithm": "file-sha256"
```

Comment:
Note: We could also name this field "serialization_xxx" instead of `hash_xxx`. Might be more consistent with the codebase (we have distinct hash and serialization code), but more confusing since intoto does not this concept. On the other hand, I recall Hayden saying he'd like to see cosign and other tools have a dedicated serialization layer, iirc

Note 2: In SLSA we typically use camel case for all field names.

Those are just fyis, feel free to ignore

Reply:
So, `serializationAlgorithm` instead of `actual_hash_algorithm`? I'll do that after the set lands, otherwise due to having to regenerate goldens there will be a ton of conflicts.